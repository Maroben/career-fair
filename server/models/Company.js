const mongoose = require("mongoose")
const Joi = require("@hapi/joi")

const companySchema = new mongoose.Schema({
	name: {
		type: String,
		required: true,
		minlength: 2,
		maxlength: 32
	},
	location: {
		type: String,
		maxlength: 16
	},
	info: {
		type: String,
		maxlength: 512
	},
	description: {
		type: String,
		maxlength: 2056
	},
	subjects: {
		type: Array
	},
	employments: {
		type: Array
	},
	tags: {
		type: Array
	},
	links: {
		type: Object
	}
})

const companyJoi = {
	name: Joi.string()
		.min(2)
		.max(32)
		.required(),
	info: Joi.string().max(512),
	description: Joi.string().max(2056),
	subjects: Joi.array(),
	employments: Joi.array(),
	tags: Joi.array(),
	links: Joi.object()
}

const locationJoi = {
	location: Joi.string().max(16)
}

module.exports = {
	Model: mongoose.model("companies", companySchema),
	attr: [
		["name", "location", "info", "description", "subjects", "employments", "links", "tags"],
		["_id", "name", "location", "info", "subjects", "tags"]
	],
	validateCompany: (company) => {
		return Joi.validate(company, companyJoi)
	},
	validateLocation: (location) => {
		return Joi.validate(location, locationJoi)
	}
}
