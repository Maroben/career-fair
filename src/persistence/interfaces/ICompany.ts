import { Document } from "mongoose"

export default interface ICompany extends Document {
    name: string
    info: string
    description: string
    location: string
}
