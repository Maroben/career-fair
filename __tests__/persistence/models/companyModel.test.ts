import _ from "lodash"
import Company, { validate, properties } from "../../../src/persistence/models/CompanyModel"
import ICompany from "../../../src/persistence/interfaces/ICompany"

describe("Company Model", () => {
    describe("Valid Company Creations", () => {
        it("Simple Company", () => {
            const company: ICompany = new Company({
                name: "Name",
                info: "Info",
                description: "Description",
                location: "Location"
            })
            expect(_.pick(company, properties)).toMatchSnapshot()
            expect(async () => await validate(company)).resolves
        })
        it("Minimal Company", () => {
            const company: ICompany = new Company({
                name: "Name",
                info: "",
                description: "",
                location: ""
            })
            expect(_.pick(company, properties)).toMatchSnapshot()
            expect(async () => await validate(company)).resolves
        })
    })

    describe("Joi Company Validation Rejections", () => {
        it("Empty Object", () => {
            const company: ICompany = new Company({})
            expect(async () => await validate(company)).rejects
        })
        it("Empty Name", () => {
            const company: ICompany = new Company({
                name: "",
                info: "Info",
                description: "Description",
                location: "Location"
            })
            expect(async () => await validate(company)).rejects
        })
        it("Too long Name", () => {
            const company: ICompany = new Company({
                name: "My Company Name is wayyyyyy too Long",
                info: "Info",
                description: "Description",
                location: "Location"
            })
            expect(async () => await validate(company)).rejects
        })
        it("Wrong data type", () => {
            const company: ICompany = new Company({
                name: 8,
                info: {},
                description: [],
                location: false
            })
            expect(async () => await validate(company)).rejects
        })
    })
})
