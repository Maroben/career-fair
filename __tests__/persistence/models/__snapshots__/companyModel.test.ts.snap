// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company Model Valid Company Creations Minimal Company 1`] = `
Object {
  "description": "",
  "info": "",
  "location": "",
  "name": "Name",
}
`;

exports[`Company Model Valid Company Creations Simple Company 1`] = `
Object {
  "description": "Description",
  "info": "Info",
  "location": "Location",
  "name": "Name",
}
`;
